// @fidanjs/integration-react/build/index.mjs
import * as React from 'react';
import { createElement, Component, Fragment } from 'react';
import { render } from 'react-dom';
import { compute } from '@fidanjs/runtime';

var reactToDom = function (Tag, attributes) {
  var childs = [], len = arguments.length - 2;
  while ( len-- > 0 ) childs[ len ] = arguments[ len + 2 ];

  var parent = document.createElement("template");
  var RenderWrapper = /*@__PURE__*/(function (superclass) {
    function RenderWrapper(props) {
      superclass.call(this, props);
      this.state = {};
      this.dynamics = [];

      for (var key in props) {
        if (typeof props[key] === "function" && props[key].hasOwnProperty("$val")) {
          this.dynamics.push(props[key]);
          this.state[key] = props[key]();
        } else { this.state[key] = props[key]; }
      }
    }

    if ( superclass ) RenderWrapper.__proto__ = superclass;
    RenderWrapper.prototype = Object.create( superclass && superclass.prototype );
    RenderWrapper.prototype.constructor = RenderWrapper;

    RenderWrapper.prototype.componentDidMount = function componentDidMount () {
      var this$1 = this;

      var props = this.props;
      compute(function () {
        var newState = {};

        for (var key in this$1.props) {
          if (typeof props[key] === "function" && props[key].hasOwnProperty("$val")) {
            newState[key] = props[key]();
          }
        }

        this$1.setState(newState);
      }, this.dynamics);
    };

    RenderWrapper.prototype.render = function render$$1 () {
      return createElement.apply(React, [ Tag, this.state ].concat( childs ));
    };

    return RenderWrapper;
  }(Component));
  render(createElement(RenderWrapper, attributes), parent);
  return parent.firstElementChild;
};
var domToReact = function (dom) {
  var integrationDomRef = function (element) {
    if (element) {
      element.replaceWith(dom.cloneNode(true));
    }
  };

  return createElement(Fragment, {}, createElement("template", {
    ref: integrationDomRef
  }));
};

export { reactToDom, domToReact };
